### PR TITLE
feature: added first user as superadmin

### PR DESCRIPTION
**Pull request by eliot-blankenberg**
Signed-off-by: eliot-blankenberg <eblankenberg@openware.center>

### CODE REVIEW

File: api/asymmetric_login_test.go [hunk 1]
```diff
@@ -120,6 +120,116 @@ func (ts *SignupTestSuite) TestWrongKeyFormatGetChallengeToken() {
...
+	user, key, err := models.FindUserWithAsymmetrickey(ts.API.db, "0x6BE46d7D863666546b77951D5dfffcF075F36E68")
+	require.NoError(ts.T(), err)
+	require.NotEmpty(ts.T(), user)
+	require.NotEmpty(ts.T(), key)
+	assert.Equal(ts.T(), key.ChallengeToken.String(), jsonData.ChallengeToken)
+	assert.Equal(ts.T(), user.Role, "authenticated")
+	assert.Equal(ts.T(), user.IsSuperAdmin, false)
+}
+
+func (ts *SignupTestSuite) TestNotFirstSignIn() {
```
Comment (dpatsora) [inline api/asymmetric_login_test.go:193]: Not clear if the feature is enabled in the test
```
ts.Config.FirstUserSuperAdmin = true
```

File: api/asymmetric_signin.go [hunk 2]
```diff
@@ -65,6 +71,15 @@ func (a *API) GetChallengeToken(w http.ResponseWriter, r *http.Request) error {
...
 				Aud:      aud,
 			})
+
+			if terr != nil {
+				return terr
+			}
+
+			if config.FirstUserSuperAdmin && first {
+				terr = user.SetSuperAdmin(tx)
+			}
```
Comment (dpatsora) [inline api/asymmetric_signin.go]: I would move this logic into (a *API) signupNewUser


```
	err = conn.Transaction(func(tx *storage.Connection) error {
		var terr error
		if terr = tx.Create(user); terr != nil {
			return internalServerError("Database error saving new user").WithInternalError(terr)
		}
		if config.FirstUserSuperAdmin && first {
				terr = user.SetRole(tx, "superadmin")
			} else {
			terr = user.SetRole(tx, config.JWT.DefaultGroupName); terr != nil {
			return internalServerError("Database error updating user").WithInternalError(terr)
		}
			}
	
		if terr = triggerEventHooks(ctx, tx, ValidateEvent, user, instanceID, config); terr != nil {
			return terr
		}
		return nil
	})
	if err != nil {
		return nil, err
	}

	return user, nil
```

Comment (dpatsora) [reply api/asymmetric_signin.go]: smth like this

Also models.IsFirstUser is a bit misleading here, as at this moment we don't have user.
I would rename to models.AnyUser or smth like this.

**Review by dpatsora — CHANGES_REQUESTED** [review]
In general, well done. Tests are good. Please, move the logic into a.signupNewUser, so the logic will be shared between all the endpoints.

File: models/user.go [hunk 3]
```diff
@@ -347,6 +357,18 @@ func findUser(tx *storage.Connection, query string, args ...interface{}) (*User,
 	return obj, nil
 }
 
+func AnyUser(tx *storage.Connection) (bool, error) {
```
Comment (dpatsora) [inline models/user.go:360]: AnyUser should return true if userCount != 0

**Review by dpatsora — CHANGES_REQUESTED** [review]

File: api/signup.go [hunk 4]
```diff
@@ -301,12 +301,27 @@ func (a *API) signupNewUser(ctx context.Context, conn *storage.Connection, param
 
 	err = conn.Transaction(func(tx *storage.Connection) error {
 		var terr error
+		exist_user, terr := models.AnyUser(tx)
```
Comment (dpatsora) [inline api/signup.go]: exist_user  -> userExist

File: models/user.go [hunk 5]
```diff
@@ -347,6 +357,18 @@ func findUser(tx *storage.Connection, query string, args ...interface{}) (*User,
...
+	obj := &User{}
+	err := tx.Eager().Q().First(obj)
+	if err != nil {
+		if errors.Cause(err) == sql.ErrNoRows {
+			return false, nil
+		}
+		return true, err
```
Comment (dpatsora) [inline models/user.go]: return false, err

**Review by dpatsora — CHANGES_REQUESTED** [review]

**Review by dpatsora — APPROVED** [review]

**Review by mnaichuk — APPROVED** [review]

**Pull request merged**